File: ccwap/server/models/sessions.py
"""Pydantic models for the sessions API."""

from typing import List, Optional, Dict
from pydantic import BaseModel

from ccwap.server.models.common import PaginationMeta


class SessionListItem(BaseModel):
    """Session in the list view."""
    session_id: str
    project_path: str
    project_display: Optional[str] = None
    first_timestamp: Optional[str] = None
    last_timestamp: Optional[str] = None
    duration_seconds: int = 0
    cost: float = 0.0
    turns: int = 0
    user_turns: int = 0
    tool_calls: int = 0
    errors: int = 0
    is_agent: bool = False
    cc_version: Optional[str] = None
    git_branch: Optional[str] = None
    model: Optional[str] = None


class SessionsResponse(BaseModel):
    """Paginated session list."""
    sessions: List[SessionListItem]
    pagination: PaginationMeta


class ToolCallDetail(BaseModel):
    """Tool call within a session replay turn."""
    tool_name: str
    file_path: Optional[str] = None
    success: bool = True
    error_message: Optional[str] = None
    error_category: Optional[str] = None
    loc_written: int = 0
    lines_added: int = 0
    lines_deleted: int = 0
    language: Optional[str] = None


class ReplayTurn(BaseModel):
    """A single turn in the session replay timeline."""
    uuid: str
    timestamp: Optional[str] = None
    entry_type: str
    model: Optional[str] = None
    input_tokens: int = 0
    output_tokens: int = 0
    cache_read_tokens: int = 0
    cache_write_tokens: int = 0
    thinking_chars: int = 0
    cost: float = 0.0
    cumulative_cost: float = 0.0
    stop_reason: Optional[str] = None
    is_sidechain: bool = False
    is_meta: bool = False
    user_prompt_preview: Optional[str] = None
    tool_calls: List[ToolCallDetail] = []


class SessionReplayResponse(BaseModel):
    """Full session replay data."""
    session_id: str
    project_path: str
    project_display: Optional[str] = None
    first_timestamp: Optional[str] = None
    last_timestamp: Optional[str] = None
    duration_seconds: int = 0
    cc_version: Optional[str] = None
    git_branch: Optional[str] = None
    is_agent: bool = False
    total_cost: float = 0.0
    total_turns: int = 0
    total_user_turns: int = 0
    total_tool_calls: int = 0
    total_errors: int = 0
    cost_by_model: Dict[str, float] = {}
    tool_distribution: Dict[str, int] = {}
    turns: List[ReplayTurn] = []

File: ccwap/frontend/src/api/sessions.ts
import { useQuery } from '@tanstack/react-query'
import { apiFetch, buildQuery } from './client'
import { sessionKeys } from './keys'
import type { DateRange } from '@/hooks/useDateRange'

interface SessionListItem {
  session_id: string
  project_path: string
  project_display: string | null
  first_timestamp: string | null
  last_timestamp: string | null
  duration_seconds: number
  cost: number
  turns: number
  user_turns: number
  tool_calls: number
  errors: number
  is_agent: boolean
  cc_version: string | null
  git_branch: string | null
  model: string | null
}

interface SessionsResponse {
  sessions: SessionListItem[]
  pagination: {
    page: number
    limit: number
    total_count: number
    total_pages: number
  }
}

export interface ToolCallDetail {
  tool_name: string
  file_path: string | null
  success: boolean
  error_message: string | null
  error_category: string | null
  loc_written: number
  lines_added: number
  lines_deleted: number
  language: string | null
}

export interface ReplayTurn {
  uuid: string
  timestamp: string | null
  entry_type: string
  model: string | null
  input_tokens: number
  output_tokens: number
  cache_read_tokens: number
  cache_write_tokens: number
  thinking_chars: number
  cost: number
  cumulative_cost: number
  stop_reason: string | null
  is_sidechain: boolean
  is_meta: boolean
  user_prompt_preview: string | null
  tool_calls: ToolCallDetail[]
}

export interface SessionReplayData {
  session_id: string
  project_path: string
  project_display: string | null
  first_timestamp: string | null
  last_timestamp: string | null
  duration_seconds: number
  cc_version: string | null
  git_branch: string | null
  is_agent: boolean
  total_cost: number
  total_turns: number
  total_user_turns: number
  total_tool_calls: number
  total_errors: number
  cost_by_model: Record<string, number>
  tool_distribution: Record<string, number>
  turns: ReplayTurn[]
}

export function useSessions(dateRange: DateRange, project?: string, page = 1) {
  return useQuery({
    queryKey: sessionKeys.list(dateRange, project, page),
    queryFn: () => apiFetch<SessionsResponse>(
      `/sessions${buildQuery({ from: dateRange.from, to: dateRange.to, project, page })}`
    ),
  })
}

export function useSessionReplay(sessionId: string) {
  return useQuery({
    queryKey: sessionKeys.replay(sessionId),
    queryFn: () => apiFetch<SessionReplayData>(`/sessions/${sessionId}/replay`),
    enabled: !!sessionId,
  })
}
